[ManagedMSE] onbufferedchange fired too many times if GPU process isn't in use.

jyavenard opened the pull request:
#### dab28312bc80bc475196b7d6221b060347254e33
<pre>
[ManagedMSE] onbufferedchange fired too many times if GPU process isn&apos;t in use.
<a href="https://bugs.webkit.org/show_bug.cgi?id=257057">https://bugs.webkit.org/show_bug.cgi?id=257057</a>
rdar://problem/109586019

Reviewed by Youenn Fablet.

In bug 253560, changes were made so that we correctly re-calculated the
buffered range and notify of the changes back to the SourcebufferPrivateClient.
However it left an issue open: the SourceBufferPrivateClient would be notified
twice whenever the buffered attribute changed.
When running in the GPU process, this wasn&apos;t a problem as the RemoteSourceBufferProxy
would then only send a single message to the content process, and so the
event was only fired once.
But in WK1, both would lead the `bufferedchanged` event to be fired, and
the test in place only ensured that it was fired, not how many times.
We remove the redundant call to `sourceBufferPrivateBufferedChanged`.

Fly-by fix: The SourceBuffer once closed would notify the SourceBufferPrivate
that it should recalculate the buffered range to execute step 6.2.2.4.2
(<a href="https://w3c.github.io/media-source/#htmlmediaelement-extensions-buffered)">https://w3c.github.io/media-source/#htmlmediaelement-extensions-buffered)</a>
which is a synchronous process.
This caused the need for UpdateBufferedFromTrackBuffers IPC message to be
synchronous as well as it required access to the SourceBuffer&apos;s track buffers
living in the GPU process.
To get around this, we add a new IPC method that allows the CP&apos;s SourceBufferPrivateRemote
to cache its own copy of all the track buffers ranges. So that when the SourceBuffer
gets ended, we can immediately locally calculate the new buffered range
without the need to query the GPU process and the sync call can be removed.
The track buffers&apos; ranges are updated in the CP&apos;s SourceBufferPrivateRemote
when:
- during the prepare append algorithm
- after an append operation
- after a remove operation
- if ManagedSourceBuffer, after a memory pressure signal.

There is now only one SourceBufferPrivate synchronous IPC method left: EvictCodedFrames.

* LayoutTests/media/media-source/media-managedmse-bufferedchange-expected.txt:
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html: log whenever
the bufferedchange event is called.
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::readyStateChanged):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::trackBuffersRanges const):
(WebCore::SourceBufferPrivate::clientReadyStateChanged):
(WebCore::SourceBufferPrivate::updateBufferedFromTrackBuffers):
(WebCore::SourceBufferPrivate::processAppendCompletedOperation):
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::removeCodedFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
(WebCore::SourceBufferPrivate::setShouldGenerateTimestamps):
(WebCore::SourceBufferPrivate::updateBufferedFromTrackBuffers):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
(WebCore::SourceBufferPrivateClient::sourceBufferPrivateTrackBuffersChanged):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateTrackBuffersChanged):
(WebKit::RemoteSourceBufferProxy::clientReadyStateChanged):
(WebKit::RemoteSourceBufferProxy::updateBufferedFromTrackBuffers): Deleted.
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::clientReadyStateChanged):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateTrackBuffersChanged):
(WebKit::SourceBufferPrivateRemote::updateBufferedFromTrackBuffers): Deleted.
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: <a href="https://commits.webkit.org/264467@main">https://commits.webkit.org/264467@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc807f154b6e487794a0f4ce33cfef55e931f2ec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7715 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9355 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/7873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7721 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9975 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7906 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10741 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7849 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8965 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9463 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6275 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7024 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14687 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/7148 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10553 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7638 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6966 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6934 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1840 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11176 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7372 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -108,36 +108,49 @@ void SourceBufferPrivate::setBufferedRanges(PlatformTimeRanges&& timeRanges, Com
+{
+    Vector<PlatformTimeRanges> trackBuffers { m_trackBufferMap.size() };
+    for (auto& trackBuffer : m_trackBufferMap.values())
+        trackBuffers.append(trackBuffer->buffered());
```
We usually do reserveInitialCapacity/uncheckedAppend.
You might be able to use WTF::map here.

Side question: are we often recomputing the same vector so that it becomes useful to cache it?

jyavenard replied:
hhhm, that's a good idea, but it would be quite involved to update the base class cache when any of the derived class modify it

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -108,36 +108,49 @@ void SourceBufferPrivate::setBufferedRanges(PlatformTimeRanges&& timeRanges, Com
 
         // 4.1 Let track ranges equal the track buffer ranges for the current track buffer.
-        PlatformTimeRanges trackRanges = trackBuffer->buffered();
+        PlatformTimeRanges trackRanges = trackBuffer;
```
auto

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -260,10 +275,12 @@ void SourceBufferPrivate::clearTrackBuffers(bool shouldReportToClient)
+        m_client->sourceBufferPrivateTrackBuffersChanged({ });
         m_client->sourceBufferPrivateReportExtraMemoryCost(totalTrackBufferSizeInBytes());
+    }
+    updateBufferedFromTrackBuffers({ }, true);
```
Preexisting, but it might be good to use a temp variable isEnded = true to improve readability.

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -458,12 +475,13 @@ void SourceBufferPrivate::removeCodedFrames(const MediaTime& start, const MediaT
-        if (!trackBuffer.removeCodedFrames(start, end, currentTime))
-            continue;
+        trackBuffer.removeCodedFrames(start, end, currentTime);
+        trackBuffers.append(trackBuffer.buffered());
```
Ditto for WTF::map here or uncheckedAppend.

youennf commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
```diff
@@ -73,13 +73,15 @@ class RemoteSourceBufferProxy final
     void sourceBufferPrivateReportExtraMemoryCost(uint64_t extraMemory) final;
-    bool isAsync() const { return true; }
+    bool isAsync() const final { return true; }
+
```
Line not needed

youennf commented on Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
```diff
@@ -185,20 +185,11 @@ void SourceBufferPrivateRemote::setMode(SourceBufferAppendMode mode)
-    setBufferedRanges(WTFMove(buffered));
+    if (isGPURunning())
+        m_gpuProcessConnection->connection().send(Messages::RemoteSourceBufferProxy::ClientReadyStateChanged(sourceIsEnded), m_remoteSourceBufferIdentifier);
+    SourceBufferPrivate::updateBufferedFromTrackBuffers(m_trackBufferRanges, sourceIsEnded);
```
This is a slight change of behaviour: do we care to call this if gpu process is not running?

jyavenard replied:
probably not, but it would give a better answer than not. though it probably no longer matters at this stage

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.h:
```diff
@@ -137,6 +137,8 @@ class SourceBufferPrivate
 #endif
 
 protected:
+    WEBCORE_EXPORT virtual void updateBufferedFromTrackBuffers(const Vector<PlatformTimeRanges>&, bool sourceIsEnded, CompletionHandler<void()>&& = [] { });
```
No longer needs to be virtual?

jyavenard replied:
indeed. good spotting

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.h:
```diff
@@ -81,6 +81,7 @@ class SourceBufferPrivate
     virtual void removedFromMediaSource() = 0;
     virtual MediaPlayer::ReadyState readyState() const = 0;
     virtual void setReadyState(MediaPlayer::ReadyState) = 0;
+    WEBCORE_EXPORT virtual void clientReadyStateChanged(bool);
```
It is a bit odd to pass a bool to clientReadyStateChanged, maybe add the name of the bool to ease readability.

jyavenard committed:
[ManagedMSE] onbufferedchange fired too many times if GPU process isn't in use.
https://bugs.webkit.org/show_bug.cgi?id=257057
rdar://problem/109586019

Reviewed by Youenn Fablet.

In bug 253560, changes were made so that we correctly re-calculated the
buffered range and notify of the changes back to the SourcebufferPrivateClient.
However it left an issue open: the SourceBufferPrivateClient would be notified
twice whenever the buffered attribute changed.
When running in the GPU process, this wasn't a problem as the RemoteSourceBufferProxy
would then only send a single message to the content process, and so the
event was only fired once.
But in WK1, both would lead the `bufferedchanged` event to be fired, and
the test in place only ensured that it was fired, not how many times.
We remove the redundant call to `sourceBufferPrivateBufferedChanged`.

Fly-by fix: The SourceBuffer once closed would notify the SourceBufferPrivate
that it should recalculate the buffered range to execute step 6.2.2.4.2
(https://w3c.github.io/media-source/#htmlmediaelement-extensions-buffered)
which is a synchronous process.
This caused the need for UpdateBufferedFromTrackBuffers IPC message to be
synchronous as well as it required access to the SourceBuffer's track buffers
living in the GPU process.
To get around this, we add a new IPC method that allows the CP's SourceBufferPrivateRemote
to cache its own copy of all the track buffers ranges. So that when the SourceBuffer
gets ended, we can immediately locally calculate the new buffered range
without the need to query the GPU process and the sync call can be removed.
The track buffers' ranges are updated in the CP's SourceBufferPrivateRemote
when:
- during the prepare append algorithm
- after an append operation
- after a remove operation
- if ManagedSourceBuffer, after a memory pressure signal.

There is now only one SourceBufferPrivate synchronous IPC method left: EvictCodedFrames.

* LayoutTests/media/media-source/media-managedmse-bufferedchange-expected.txt:
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html: log whenever
the bufferedchange event is called.
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::readyStateChanged):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::trackBuffersRanges const):
(WebCore::SourceBufferPrivate::clientReadyStateChanged):
(WebCore::SourceBufferPrivate::updateBufferedFromTrackBuffers):
(WebCore::SourceBufferPrivate::processAppendCompletedOperation):
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::removeCodedFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
(WebCore::SourceBufferPrivate::setShouldGenerateTimestamps):
(WebCore::SourceBufferPrivate::updateBufferedFromTrackBuffers):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
(WebCore::SourceBufferPrivateClient::sourceBufferPrivateTrackBuffersChanged):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateTrackBuffersChanged):
(WebKit::RemoteSourceBufferProxy::clientReadyStateChanged):
(WebKit::RemoteSourceBufferProxy::updateBufferedFromTrackBuffers): Deleted.
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::clientReadyStateChanged):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateTrackBuffersChanged):
(WebKit::SourceBufferPrivateRemote::updateBufferedFromTrackBuffers): Deleted.
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: https://commits.webkit.org/264467@main
